util/flop_latency.py

```diff
@@ -0,0 +1,440 @@
+    ctx.allow_unregistered_dialects = True
+    mlir_module = Module.parse(mlir_module_str, ctx)
+
+    traverse_op_region_block_iterators(mlir_module.operation, print_op)
+    return total_latency
+
+
+def count_flop_latency_in_mlir_llvm_module(node_name, mlir_module_str):
+    total_latency = 0
+    os.makedirs("/tmp/mlir_llvm_modules", exist_ok=True)
+    mlir_module_fp = f"/tmp/mlir_llvm_modules/{node_name}.llvm.mlir"
+    with open(mlir_module_fp, "w") as mlir_file:
+        mlir_file.write(mlir_module_str)
+
+    mlir_translate_path = os.environ["MLIR_TRANSLATE_PATH"]
+    llvm_ir_fp = mlir_module_fp.replace(".mlir", "")
+    subprocess.run(
+        [mlir_translate_path, "--mlir-to-llvmir", mlir_module_fp, "-o", llvm_ir_fp]
+    )
+
+    opt_path = os.environ["OPT_PATH"]
+    unrolled_llvm_ir_fp = llvm_ir_fp.replace(".llvm", ".unrolled.llvm")
+    subprocess.run(
+        [
+            opt_path,
+            "-S",
+            "--indvars",
+            "--loop-unroll",
+            "--unroll-count=1000000000",
+            llvm_ir_fp,
+            "-o",
+            unrolled_llvm_ir_fp,
+        ]
+    )
+
+    reg = re.compile(f"(fcmp|fadd|fsub|fmin|fmax|fmul|fdiv)")
+    with open(unrolled_llvm_ir_fp) as unrolled_llvm_ir:
+        res = reg.findall(unrolled_llvm_ir.read())
+    op_counts = Counter(res)
+    for op, count in op_counts.items():
+        total_latency += LLVM_FP_OP_LATENCIES[op] * count
+
+    return node_name, total_latency
+
+
+def count_flop_latency_in_mlir_modules_using_mlir(
+    node_mlir_module_strs: List[Tuple[str, str]]
+) -> dict[str, int]:
+    with Pool(cpu_count() - 1) as pool:
+        latencies = pool.map(
+            lambda x: (x[0], count_flop_latency_in_mlir_module(x[1])),
+            node_mlir_module_strs,
+        )
+    return dict(latencies)
+
+
+def count_flop_latency_in_mlir_modules_using_llvm(
+    node_mlir_module_strs: List[Tuple[str, str]]
+) -> dict[str, int]:
+    with Pool(cpu_count() - 1) as pool:
+        latencies = pool.map(
+            lambda x: count_flop_latency_in_mlir_llvm_module(*x), node_mlir_module_strs
+        )
+    return dict(latencies)
+
+
+def compute_graph_slice_traffic(nodes, node_shapes):
+    # Collect incoming edges for every slice of the graph in order to
+    # compute ingress and egress data (bytes) further down.
+    cumu_inputs = []
+    for node in nodes:
+        if cumu_inputs:
+            cumu_inputs.append(
+                set((n.name, node.name) for n in node.all_input_nodes) | cumu_inputs[-1]
+            )
+        else:
+            cumu_inputs.append(set((n.name, node.name) for n in node.all_input_nodes))
+
+    # Collect outgoing edges for every slice of the graph in order to
+    # compute ingress and egress data (bytes) further down.
+    cumu_outputs = []
+    for node in reversed(nodes):
+        if cumu_outputs:
+            cumu_outputs.append(
+                set((node.name, n.name) for n in node.users) | cumu_outputs[-1]
+            )
```

Comment:
"prefix sums" for edges, i.e. an increasing collection of out-edges - **note that this loop traverses the node list in reverse**.